Graphite TCP connection problems

After some time the program couldn't maintain Graphite connection anymore and [marpaia/graphite-golang](https://github.com/marpaia/graphite-golang) library which is used here doesn't re-establish it:
```2019/10/03 08:13:43 [ERROR] Can't send metrics to Graphite, write tcp *.*.*.*:57224->*.*.*.*:2003: write: broken pipe```
Script need to either die in such a case (so service would be restarted in case of running in Kubernetes) or be rewritten to properly handle such situation.

--- a/go.mod
+++ b/go.mod
@@ -7,7 +7,7 @@ require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/google/go-cmp v0.3.1 // indirect
 	github.com/jessevdk/go-flags v1.4.0
-	github.com/marpaia/graphite-golang v0.0.0-20190519024811-caf161d2c2b1
+	github.com/jtaczanowski/go-graphite-client v0.1.1
 	github.com/pkg/errors v0.8.1
 	github.com/stretchr/testify v1.4.0
 	go.opencensus.io v0.22.1 // indirect

--- a/graphite/graphite.go
+++ b/graphite/graphite.go
@@ -15,33 +15,32 @@
 package graphite
 
 import (
-	"fmt"
-	"strconv"
 	"time"
 
 	"github.com/bookingcom/cloudsec-metrics/api"
-	"github.com/marpaia/graphite-golang"
 )
 
 // GenerateComplianceInfo returns metrics from given compliance info
-func GenerateComplianceInfo(timeNow int64, prefix string, ci []api.ComplianceInfo) []graphite.Metric {
-	var metrics []graphite.Metric
+func GenerateComplianceInfo(prefix string, ci []api.ComplianceInfo) []map[string]float64 {
+	var metrics []map[string]float64
 	for _, entry := range ci {
 		metricPrefix := prefix + escapeMetricName(entry.Name)
-		metrics = append(metrics, graphite.NewMetric(metricPrefix+".policies_total", strconv.Itoa(entry.PoliciesCount), timeNow))
-		metrics = append(metrics, graphite.NewMetric(metricPrefix+".assets_passed", strconv.Itoa(entry.PassedAssetsCount), timeNow))
-		metrics = append(metrics, graphite.NewMetric(metricPrefix+".assets_failed", strconv.Itoa(entry.FailedAssetsCount), timeNow))
-		metrics = append(metrics, graphite.NewMetric(metricPrefix+".assets_total", strconv.Itoa(entry.TotalAssetsCount), timeNow))
+		metrics = append(metrics, map[string]float64{
+			metricPrefix + ".policies_total": float64(entry.PoliciesCount),
+			metricPrefix + ".assets_passed":  float64(entry.PassedAssetsCount),
+			metricPrefix + ".assets_failed":  float64(entry.FailedAssetsCount),
+			metricPrefix + ".assets_total":   float64(entry.TotalAssetsCount),
+		})
 	}
 	return metrics
 }
 
 // GenerateSSCSourcesDelay returns metrics from given delay map
-func GenerateSSCSourcesDelay(timeNow int64, prefix string, delay map[string]time.Duration) []graphite.Metric {
-	var metrics []graphite.Metric
+func GenerateSSCSourcesDelay(prefix string, delay map[string]time.Duration) []map[string]float64 {
+	var metrics []map[string]float64
 	for name, duration := range delay {
 		metricPrefix := prefix + escapeMetricName(name)
-		metrics = append(metrics, graphite.NewMetric(metricPrefix+".seconds", fmt.Sprintf("%f", duration.Seconds()), timeNow))
+		metrics = append(metrics, map[string]float64{metricPrefix + ".seconds": duration.Seconds()})
 	}
 	return metrics
 }

--- a/graphite/graphite_test.go
+++ b/graphite/graphite_test.go
@@ -18,32 +18,33 @@ import (
 	"testing"
 	"time"
 
-	"github.com/bookingcom/cloudsec-metrics/api"
-	"github.com/marpaia/graphite-golang"
 	"github.com/stretchr/testify/assert"
+
+	"github.com/bookingcom/cloudsec-metrics/api"
 )
 
 func TestGraphite(t *testing.T) {
-	assert.Nil(t, GenerateComplianceInfo(0, "", nil),
+	assert.Nil(t, GenerateComplianceInfo("", nil),
 		"Run with no metrics should return nil")
 	assert.Equal(t,
-		[]graphite.Metric{
-			{Name: "test_name_.policies_total", Value: "1", Timestamp: 0},
-			{Name: "test_name_.assets_passed", Value: "2", Timestamp: 0},
-			{Name: "test_name_.assets_failed", Value: "3", Timestamp: 0},
-			{Name: "test_name_.assets_total", Value: "4", Timestamp: 0}},
-		GenerateComplianceInfo(0, "", []api.ComplianceInfo{
+		[]map[string]float64{{
+			"test_name_.policies_total": 1,
+			"test_name_.assets_passed":  2,
+			"test_name_.assets_failed":  3,
+			"test_name_.assets_total":   4,
+		}},
+		GenerateComplianceInfo("", []api.ComplianceInfo{
 			{Name: "test{name}",
 				PoliciesCount:     1,
 				PassedAssetsCount: 2,
 				FailedAssetsCount: 3,
 				TotalAssetsCount:  4}}),
 		"Single metric send to empty Graphite should do nothing and return no errors")
-	assert.Nil(t, GenerateSSCSourcesDelay(0, "", nil),
+	assert.Nil(t, GenerateSSCSourcesDelay("", nil),
 		"Run with no metrics should return nil")
 	assert.Equal(t,
-		[]graphite.Metric{{Name: "test.seconds", Value: "0.065000", Timestamp: 0}},
-		GenerateSSCSourcesDelay(0, "", map[string]time.Duration{"test": time.Millisecond * 65}),
+		[]map[string]float64{{"test.seconds": 0.065}},
+		GenerateSSCSourcesDelay("", map[string]time.Duration{"test": time.Millisecond * 65}),
 		"Single metric send to empty Graphite should do nothing and return no errors")
 	assert.Equal(t, "_test_of_metric", escapeMetricName("(test)of/metric"))
 }

--- a/main.go
+++ b/main.go
@@ -17,14 +17,14 @@ package main
 import (
 	"log"
 	"os"
-	"strconv"
 	"time"
+
+	"github.com/jessevdk/go-flags"
+	g "github.com/jtaczanowski/go-graphite-client"
+	"github.com/pkg/errors"
 
 	"github.com/bookingcom/cloudsec-metrics/api"
 	"github.com/bookingcom/cloudsec-metrics/graphite"
-	"github.com/jessevdk/go-flags"
-	g "github.com/marpaia/graphite-golang"
-	"github.com/pkg/errors"
 )
 
 type opts struct {
@@ -50,7 +50,7 @@ type collectors struct {
 }
 
 type senders struct {
-	graphite *g.Graphite
+	graphite *g.Client
 }
 
 type metrics struct {
@@ -76,10 +76,7 @@ func main() {
 	if err != nil {
 		log.Fatalf("[ERROR] Can't initialise collectors, %v", err)
 	}
-	senders, err := prepareSenders(opts)
-	if err != nil {
-		log.Fatalf("[ERROR] Can't initialise senders, %v", err)
-	}
+	senders := prepareSenders(opts)
 
 	for ticker := time.NewTicker(opts.CollectPeriod); true; <-ticker.C {
 		collectMetrics(metrics, collectors, "https://status.cloud.google.com/incidents.json")
@@ -107,16 +104,12 @@ func prepareCollectors(opts opts) (*collectors, error) {
 }
 
 // create and return a pointer to senders
-func prepareSenders(opts opts) (*senders, error) {
-	var err error
+func prepareSenders(opts opts) *senders {
 	var senders = &senders{}
 	if opts.GraphiteHost != "" {
-		if senders.graphite, err = g.GraphiteFactory("tcp", opts.GraphiteHost, opts.GraphitePort, opts.GraphitePrefix); err != nil {
-			return nil, errors.Wrap(err, "can't create Graphite")
-		}
-		senders.graphite.DisableLog = true
+		senders.graphite = g.NewClient(opts.GraphiteHost, opts.GraphitePort, opts.GraphitePrefix, "tcp")
 	}
-	return senders, nil
+	return senders
 }
 
 // collectMetrics collects metrics into referenced metrics object using provided collectors
@@ -141,15 +134,14 @@ func collectMetrics(metrics *metrics, collectors *collectors, googleHealthDashbo
 // sendMetrics sends metrics to initialised senders
 func sendMetrics(metrics *metrics, senders *senders, opts opts) {
 	if senders.graphite != nil {
-		var graphiteMetrics []g.Metric
-		timeNow := time.Now().Unix()
+		var graphiteMetrics []map[string]float64
 		if metrics.complianceInfo != nil {
-			graphiteMetrics = append(graphiteMetrics, graphite.GenerateComplianceInfo(timeNow, opts.CompliancePrefix, metrics.complianceInfo)...)
-			graphiteMetrics = append(graphiteMetrics, g.NewMetric(opts.PrismaHealthMetricName, strconv.Itoa(metrics.prismaHealthStatus), timeNow))
+			graphiteMetrics = append(graphiteMetrics, graphite.GenerateComplianceInfo(opts.CompliancePrefix, metrics.complianceInfo)...)
+			graphiteMetrics = append(graphiteMetrics, map[string]float64{opts.PrismaHealthMetricName: float64(metrics.prismaHealthStatus)})
 		}
-		graphiteMetrics = append(graphiteMetrics, graphite.GenerateSSCSourcesDelay(timeNow, opts.SCCDelayPrefix, metrics.googleSourcesDelay)...)
-		graphiteMetrics = append(graphiteMetrics, g.NewMetric(opts.SCCHealthMetricName, strconv.Itoa(metrics.googleSCCHealthStatus), timeNow))
-		if err := senders.graphite.SendMetrics(graphiteMetrics); err != nil {
+		graphiteMetrics = append(graphiteMetrics, graphite.GenerateSSCSourcesDelay(opts.SCCDelayPrefix, metrics.googleSourcesDelay)...)
+		graphiteMetrics = append(graphiteMetrics, map[string]float64{opts.SCCHealthMetricName: float64(metrics.googleSCCHealthStatus)})
+		if err := senders.graphite.SendData(graphiteMetrics); err != nil {
 			log.Printf("[ERROR] Can't send metrics to Graphite, %v", err)
 		}
 	}

--- a/main_test.go
+++ b/main_test.go
@@ -17,9 +17,10 @@ package main
 import (
 	"testing"
 
-	"github.com/bookingcom/cloudsec-metrics/api"
-	"github.com/marpaia/graphite-golang"
+	"github.com/jtaczanowski/go-graphite-client"
 	"github.com/stretchr/testify/assert"
+
+	"github.com/bookingcom/cloudsec-metrics/api"
 )
 
 func TestPrepareCollectors(t *testing.T) {
@@ -46,12 +47,8 @@ func TestPrepareCollectors(t *testing.T) {
 }
 
 func TestPrepareSenders(t *testing.T) {
-	s, err := prepareSenders(opts{})
-	assert.NoError(t, err)
+	s := prepareSenders(opts{})
 	assert.Equal(t, &senders{}, s, "No senders initialised without options provided")
-	s, err = prepareSenders(opts{GraphiteHost: "bad_host"})
-	assert.Error(t, err, "Bad Graphite hostname results in error")
-	assert.Nil(t, s, "Bad Graphite hostname results in no senders created")
 }
 
 func TestCollectMetrics(t *testing.T) {
@@ -62,6 +59,6 @@ func TestCollectMetrics(t *testing.T) {
 
 func TestSendMetrics(t *testing.T) {
 	m := metrics{complianceInfo: []api.ComplianceInfo{}}
-	sendMetrics(&m, &senders{graphite: &graphite.Graphite{}}, opts{})
+	sendMetrics(&m, &senders{graphite: &graphite.Client{}}, opts{})
 	assert.Equal(t, metrics{complianceInfo: []api.ComplianceInfo{}}, m, "Metrics unchanged after send function call")
 }

